concordium-consensus/src/Concordium/KonsensusV1/Consensus/Blocks.hs

```diff
@@ -1371,3 +1377,28 @@ makeBlock = do
             void . onTimeout (DelayUntil . timestampToUTCTime $ blockTimestamp block) $ do
                 sendBlock block
                 checkedValidateBlock block
+
+-- | Compute the missed rounds for each validator. Starts from the given parent
+--   block up to (not including) the given final round.
+computeMissedRounds ::
+    forall m.
+    (BlockState m ~ HashedPersistentBlockState (MPV m), BlockStateQuery m) =>
+    Option TimeoutCertificate ->
+    BlockPointer (MPV m) ->
+    Round ->
+    m [(BakerId, Word16)]
+computeMissedRounds tc _parent _rnd | isAbsent tc = return []
+computeMissedRounds _tc parent rnd = do
+    let parentBlockState = bpState parent
+    validators <- getCurrentEpochBakers parentBlockState
+    seedState <- getSeedState parentBlockState
+    let missedRoundsValidators =
+            let beginRound = blockRound parent
+                endRound = rnd - 1
+                leNonce = seedState ^. currentLeadershipElectionNonce
+            in  [ winner ^. bakerIdentity
+                  | r <- [beginRound .. endRound],
+                    let winner = getLeaderFullBakers validators leNonce r
+                ]
+    let missedRounds = [(head g, fromIntegral $ length g) | g <- group missedRoundsValidators]
```

Comment:
I can't quite follow your reasoning. You would have to do the grouping after the call to `computeMissedRounds` then to compute the delta, that doesn't any better.

Reply:
You don't have to do the grouping at all. You just pass the ungrouped `missedRoundsValidators` right through and `doUpdateMissedRounds` just adds 1 for each.

Reply:
I think this is not a great idea because we would have to call `modifyBakerPoolRewardDetailsInPoolRewards` for every missed round instead of batching these calls together per validator id. `modifyBakerPoolRewardDetailsInPoolRewards`  loads a reference and does a binary search for the validator id, so this will have worse performance. I'd prefer to keep it the way it is and batch the calls for now. If we ever need the list of missed rounds we could revisit.

Reply:
Part of my problem with this is that the code doesn't quite do what I would expect. In particular, if `missedRoundsValidators = [1,2,1]` then you end up with `missedRounds = [(1,1), (2,1), (1,1)]`. So if the intention really is to batch them per validator id, then I recommend constructing a `Map BakerId Word64` instead.

Generally, looking at the definition of `missedRounds`, there are a few things that raise flags for me:
- `head` is partial, so I need to know that `g` cannot be empty. This is satisfied, but I might prefer something like `[(h, fromIntegral $ 1 + length t) | (h:t) <- group missedRoundsValidators]`.
- `fromIntegral` can be problematic, but here it's fine, since a positive `Int` will fit in a `Word64` just fine, and if we're constructing a list that is anywhere near the limit, something has already gone wrong.
- `length` on a list is O(n), but it's needed in this context, so it's fine.
- `group` is a list operation that maybe doesn't behave exactly how I might guess. We've been bitten by this sort of thing in the past - the mainnet outage was down to the fact that the `union` of sorted lists is not necessarily sorted. Here, in particular, `group` is not going to group all instances if the list is not sorted.

I would probably prefer something along the lines of:
```
computeMissedRounds ... = makeMissedRounds Map.empty (blockRound parent)
    where
        getLeader = getLeaderFullBakers validators (seedState ^. currentLeadershipElectionNonce)
        makeMissedRounds !m !r
            | r >= rnd = m
            | otherwise = makeMissedRounds (Map.insertWith (+) (getLeader r) 1 m) (r + 1)
```

Reply:
That's a fair pint, I wasn't aware of this behaviour of `group` either. I'll change it to your suggestion.

Reply:
Changed.